### ISSUE TITLE
Overriding getattr breaks inline subclassing

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#2912]
**Describe the bug**

```py
from typing import Type

class Foo:
    @classmethod
    def bar(cls) -> None:
        print('hello!')

class Models:
    def __getattr__(self, attr: str) -> Type[Foo]:
        return Foo

models = Models()

class User(models.BaseUser):  # error
    pass

User.bar()
```
```
bar.py
  bar.py:14:12 - error: Expected class type but received "Models" (reportGeneralTypeIssues)
  bar.py:14:12 - error: Base class type is unknown, obscuring type of derived class (reportUntypedBaseClass)
2 errors, 0 warnings, 0 informations
```

**Expected behavior**

Inferred subclass type should be `Foo`

**VS Code extension or command-line**

Command-line: 1.1.212

**Additional context**

Moving the attribute access to a type alias fixes the error:

```py
BaseUser = models.BaseUser

class User(BaseUser):
    pass
```

**Comment by erictraut** [comment]
Yeah, this is intended behavior. The fact that it works with a type alias is arguably a bug. Mypy doesn't allow either of these. The problem is that you're dynamically constructing a class hierarchy, and a static type checker needs to be able to understand the class hierarchy statically.

I recently looked at mirroring mypy's behavior and not allowing any type of the form `Type[X]` as a base class, but we have a lot of code that depends on the current behavior. So for now, I left it as is.

So if you switch to a type alias, this will work in pyright for now, but it won't work in mypy.

**Comment by erictraut** [comment]
Here's a way that will generally make type checkers happe:
```python
if TYPE_CHECKING:
    UserBase = Foo
else:
    UserBase = models.BaseUser

class User(UserBase):
    pass
```
